<a href="https://colab.research.google.com/github/lauren-forti/Programming-for-Data-Science/blob/main/forti_assignment6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Assignment #6**
## DSP 310
## Lauren Forti

In [ ]:
# preamble to be able to run notebooks in Jupyter and Colab
try:
    from google.colab import drive
    import sys
    
    drive.mount('/content/drive')
    notes_home = "/content/drive/Shared drives/CSC310/notes/"
    user_home = "/content/drive/My Drive/"
    
    sys.path.insert(1,notes_home) # let the notebook access the notes folder

    !pip install PyMySQL

except ModuleNotFoundError:
    notes_home = "" # running native Jupyter environment -- notes home is the same as the notebook
    user_home = ""  # under Jupyter we assume the user directory is the same as the notebook

Mounted at /content/drive
     |████████████████████████████████| 51kB 3.2MB/s 


In [ ]:
# our database credentials
host = 'testdb.cwy05wfzuxbv.us-east-1.rds.amazonaws.com'
database = 'world'
user = 'csc310'
password = 'csc310$is$fun'

# Database schema

<img src="https://static.packt-cdn.com/products/9781788390415/graphics/cac1f609-1c45-46d7-b066-d9481ceddf18.png">



# Assignment

## 1. What is the name of the country with country code 'ZAF'?

In [ ]:
# query city table
# if the following import fails try: conda install -c anaconda pymysql
import pymysql as sql
import pandas as pd

# Open database connection to our test database
db = sql.connect(host=host, 
                 user=user, 
                 password=password, 
                 database=database)

# get data using a pandas dataframe
sql_string = \
'''
SELECT
      name as Name,
      code as CountryCode
FROM
      country
WHERE
      code = 'ZAF'
'''

data = pd.read_sql(sql_string, con=db) 
print(data)

# disconnect from server
db.close()

           Name CountryCode
0  South Africa         ZAF


The country with the country code "ZAF" is **South Africa**.

## 2. How many cities are there in the US with less than 100,000 residents according to the world db? (hint: use count and also conditions can be combined with boolean operators such as 'and'/'or')

In [ ]:
# Open database connection to our test database
db = sql.connect(host=host, 
                 user=user, 
                 password=password, 
                 database=database)

# get data using a pandas dataframe
sql_string = \
'''
SELECT 
    COUNT(city.name) as Number
FROM 
    city 
JOIN 
    country 
ON 
    city.countrycode = country.code
WHERE
    city.population < 100000
AND
    country.name = 'United States'
'''

data = pd.read_sql(sql_string, con=db) 
print(data)

# disconnect from server
db.close()

   Number
0      29


There are **29** cities in the US with less than 100,000 residents according to the database.

## 3. How many different languages are there in the db?

In [ ]:
# Open database connection to our test database
db = sql.connect(host=host, 
                 user=user, 
                 password=password, 
                 database=database)

# get data using a pandas dataframe
sql_string = \
'''
SELECT 
    COUNT(DISTINCT language) as Number
FROM 
    countrylanguage
'''
data = pd.read_sql(sql_string, con=db) 
print(data)

# disconnect from server
db.close()

   Number
0     457


There are **457** languages in the database.

## 4. In which countries, according the db, is English the official language
### Bonus: Print the actual country names not just the country codes!

In [ ]:
# Open database connection to our test database
db = sql.connect(host=host, 
                 user=user, 
                 password=password, 
                 database=database)

# get data using a pandas dataframe
sql_string = \
'''
SELECT
    code as CountryCode,
    name as CountryName,
    language as Language,
    isofficial as Official
FROM 
    country 
JOIN 
    countrylanguage
ON 
    countrylanguage.countrycode = country.code
WHERE
    language = 'English'
AND
    isofficial = 'T'
'''

data = pd.read_sql(sql_string, con=db) 

print(data)
print('')

# print out country names and codes
print('Countries where English is the Official Language:')
print('--------------------------------------------------')
country_data = data['CountryName'] + ' (' + data['CountryCode'] + ')'
for item in country_data:
  print(item, end = ', ')

# disconnect from server
db.close()

   CountryCode                           CountryName Language Official
0          AIA                              Anguilla  English        T
1          ASM                        American Samoa  English        T
2          ATG                   Antigua and Barbuda  English        T
3          AUS                             Australia  English        T
4          BLZ                                Belize  English        T
5          BMU                               Bermuda  English        T
6          BRB                              Barbados  English        T
7          CAN                                Canada  English        T
8          CCK               Cocos (Keeling) Islands  English        T
9          CXR                      Christmas Island  English        T
10         CYM                        Cayman Islands  English        T
11         FLK                      Falkland Islands  English        T
12         GBR                        United Kingdom  English        T
13    

**Countries where English is the Official Language According to the Database:**

Anguilla (AIA), American Samoa (ASM), Antigua and Barbuda (ATG), Australia (AUS), Belize (BLZ), Bermuda (BMU), Barbados (BRB), Canada (CAN), Cocos (Keeling) Islands (CCK), Christmas Island (CXR), Cayman Islands (CYM), Falkland Islands (FLK), United Kingdom (GBR), Gibraltar (GIB), Guam (GUM), Hong Kong (HKG), Ireland (IRL), Saint Kitts and Nevis (KNA), Saint Lucia (LCA), Lesotho (LSO), Marshall Islands (MHL), Malta (MLT), Northern Mariana Islands (MNP), Montserrat (MSR), Norfolk Island (NFK), Niue (NIU), Nauru (NRU), New Zealand (NZL), Palau (PLW), Saint Helena (SHN), Seychelles (SYC), Turks and Caicos Islands (TCA), Tokelau (TKL), Tonga (TON), Tuvalu (TUV), United States Minor Outlying Islands (UMI), United States (USA), Saint Vincent and the Grenadines (VCT), Virgin Islands, British (VGB), Virgin Islands, U.S. (VIR), Vanuatu (VUT), Samoa (WSM), South Africa (ZAF), Zimbabwe (ZWE)

## 5. What are the different languages spoken in Angola?

In [ ]:
# Open database connection to our test database
db = sql.connect(host=host, 
                 user=user, 
                 password=password, 
                 database=database)

# get data using a pandas dataframe
sql_string = \
'''
SELECT
    name as CountryName,
    countrylanguage.language as Language
FROM 
    country 
JOIN 
    countrylanguage
ON 
    countrylanguage.countrycode = country.code
WHERE
    name = 'Angola'
'''

data = pd.read_sql(sql_string, con=db) 

print(data)
print('')
for lang in data['Language']:
  print(lang, end = ', ')

# disconnect from server
db.close()

  CountryName          Language
0      Angola              Ambo
1      Angola            Chokwe
2      Angola             Kongo
3      Angola           Luchazi
4      Angola  Luimbe-nganguela
5      Angola            Luvale
6      Angola            Mbundu
7      Angola   Nyaneka-nkhumbi
8      Angola         Ovimbundu

Ambo, Chokwe, Kongo, Luchazi, Luimbe-nganguela, Luvale, Mbundu, Nyaneka-nkhumbi, Ovimbundu, 

**Different Languages in Angola:** Ambo, Chokwe, Kongo, Luchazi, Luimbe-nganguela, Luvale, Mbundu, Nyaneka-nkhumbi, and Ovimbundu.